src/GUI/GraphPanel.java

```diff
@@ -375,30 +374,38 @@ private List<Vector2D> calculateIntersections() {
         double minT = toWorldCoordinates(new Vector2D(0, 0)).x;
         double maxT = toWorldCoordinates(new Vector2D(getWidth(), 0)).x;
         double step = 0.001;
+
         if (functions.size() < 2) {
             return intersections;
         }
 
-        for (double x = minT; x <= maxT; x += step) {
-            for (int i = 0; i < functions.size(); i++) {
-                for (int j = i + 1; j < functions.size(); j++) {
-                    double y1Prev = functions.get(i).evaluate(x - step).y;
-                    double y1Curr = functions.get(i).evaluate(x).y;
-                    double y2Prev = functions.get(j).evaluate(x - step).y;
-                    double y2Curr = functions.get(j).evaluate(x).y;
-
-                    // subtract values and calculate approximate roots => same as calcRoots
-                    if ((y1Prev - y2Prev) * (y1Curr - y2Curr) < 0) {
-                        double intersectX = x - step / 2;
-                        double intersectY = (y1Curr + y2Curr) / 2;
+        double EPSILON = 1E-8;
+
+        for (int i = 0; i < functions.size(); i++) {
+            for (int j = i + 1; j < functions.size(); j++) {
+                double prevY = functions.get(i).evaluate(minT).y - functions.get(j).evaluate(minT).y;
```

Comment:
It may be more readable to get the two functions once at the start of the inner loop instead of doing functions.get(i) everywhere

Reply:
Yeah its a valid point. Done